File: leetcode/155.min-stack.py
# @leet start
class MinStack:

    def __init__(self):
        self.dec_stack: list[int] = []
        self.stack: list[int] = []

    def push(self, val: int) -> None:
        if not self.dec_stack or val <= self.dec_stack[-1]:
            self.dec_stack.append(val)

        self.stack.append(val)

    def pop(self) -> None:
        res = self.stack.pop()

        if res == self.dec_stack[-1]:
            self.dec_stack.pop()

    def top(self) -> int:
        return self.stack[-1]

    def getMin(self) -> int:
        return self.dec_stack[-1]


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()
# @leet end

File: leetcode/155.min-stack.js
// @leet start

var MinStack = function () {
  this.stack = [];
  this.monoStack = []; // monotonically nonincreasing
};

/**
 * @param {number} value
 * @return {void}
 */
MinStack.prototype.push = function (value) {
  this.stack.push(value);

  if (
    !this.monoStack.length ||
    value <= this.monoStack[this.monoStack.length - 1]
  ) {
    this.monoStack.push(value);
  }
};

/**
 * @return {void}
 */
MinStack.prototype.pop = function () {
  const val = this.stack.pop();

  if (val === this.monoStack[this.monoStack.length - 1]) {
    this.monoStack.pop();
  }
};

/**
 * @return {number}
 */
MinStack.prototype.top = function () {
  return this.stack[this.stack.length - 1];
};

/**
 * @return {number}
 */
MinStack.prototype.getMin = function () {
  return this.monoStack[this.monoStack.length - 1];
};

/**
 * Your MinStack object will be instantiated and called as such:
 * var obj = new MinStack()
 * obj.push(value)
 * obj.pop()
 * var param_3 = obj.top()
 * var param_4 = obj.getMin()
 */
// @leet end
